internal/pkg/skuba/kubernetes/versions.go

```diff
@@ -142,27 +153,49 @@ var (
 func ComponentVersionWithAvailableVersions(component Component, clusterVersion *version.Version, availableVersions KubernetesVersions) string {
 	currentKubernetesVersion := availableVersions[clusterVersion.String()]
 	switch component {
+	case Kubelet:
+		return currentKubernetesVersion.ControlPlaneHostVersion.KubeletVersion
+	case ContainerRuntime:
+		return currentKubernetesVersion.ControlPlaneHostVersion.ContainerRuntimeVersion
 	case Hyperkube:
-		return currentKubernetesVersion.ControlPlaneComponentsVersion.HyperkubeVersion
+		return currentKubernetesVersion.ControlPlaneContainerVersion.HyperkubeVersion
 	case Etcd:
-		return currentKubernetesVersion.ControlPlaneComponentsVersion.EtcdVersion
-	case ContainerRuntime:
-		return currentKubernetesVersion.ComponentsVersion.ContainerRuntimeVersion
-	case Kubelet:
-		return currentKubernetesVersion.ComponentsVersion.KubeletVersion
+		return currentKubernetesVersion.ControlPlaneContainerVersion.EtcdVersion
 	case CoreDNS:
-		return currentKubernetesVersion.ComponentsVersion.CoreDNSVersion
+		return currentKubernetesVersion.ControlPlaneContainerVersion.CoreDNSVersion
 	case Pause:
-		return currentKubernetesVersion.ComponentsVersion.PauseVersion
+		return currentKubernetesVersion.ControlPlaneContainerVersion.PauseVersion
 	case Tooling:
-		return currentKubernetesVersion.ComponentsVersion.ToolingVersion
+		return currentKubernetesVersion.ControlPlaneContainerVersion.ToolingVersion
 	}
-	log.Fatalf("unknown component %q", component)
+	klog.Fatalf("unknown component %q", component)
 	panic("unreachable")
 }
 
 func ComponentVersionForClusterVersion(component Component, clusterVersion *version.Version) string {
-	return ComponentVersionWithAvailableVersions(component, clusterVersion, Versions)
+	return ComponentVersionWithAvailableVersions(component, clusterVersion, supportVersions)
+}
+
+func ComponentContainerImageWithAvailableVersions(component Component, clusterVersion *version.Version, availableVersions KubernetesVersions) string {
+	imageTag := ComponentVersionWithAvailableVersions(component, clusterVersion, availableVersions)
+	switch component {
```

Comment:
I have to admit that I don't fully understand why we need the `Component` type at all (it's just a `string` alias with no methods associated with it). That being said, if this is just a string, couldn't this whole switch statement be modified to be simply something like:

```go
return images.GetGenericImage(skuba.ImageRepository, component, imageTag)
```

To be honest, I'm not sure when is that "unknown component" case ever reached. Do we even expect that to happen? (I think we don't, that's why we have the `panic` there, but still, it feels weird). To counter this I'd log the image repository being picked before returning. Otherwise, I'm sure there has to be a more clever way to handle this (on the top of my head: instead of having separate variables, have a slice of possible components or a map if it's easier, then check that the given component exists on that map). Otherwise, it feels like this switch would grow indefinitely in the future (i.e. new components and such), but maybe I'm wrong.

Reply:
> I have to admit that I don't fully understand why we need the `Component` type at all (it's just a `string` alias with no methods associated with it). That being said, if this is just a string, couldn't this whole switch statement be modified to be simply something like:
> 
> ```go
> return images.GetGenericImage(skuba.ImageRepository, component, imageTag)
> ```
Because we have a container image called `skuba-tooling`, the variable cannot be `skuba-tooling` :)

> To be honest, I'm not sure when is that "unknown component" case ever reached. Do we even expect that to happen? (I think we don't, that's why we have the `panic` there, but still, it feels weird). To counter this I'd log the image repository being picked before returning. Otherwise, I'm sure there has to be a more clever way to handle this (on the top of my head: instead of having separate variables, have a slice of possible components or a map if it's easier, then check that the given component exists on that map). Otherwise, it feels like this switch would grow indefinitely in the future (i.e. new components and such), but maybe I'm wrong.

You are right, this is my general thought before, maybe it's better than current.
In this way, we could cover one case that the control plane container image name might be changed in the future. Also, no need to manually add a new component to the switch case. 
 WDYT?
```
type ContainerImageTag struct {
	Name string
	Tag  string
}
type ControlPlaneHostVersion struct {
	KubeletVersion          string
	ContainerRuntimeVersion string
}
type ControlPlaneContainerVersion map[Component]*ContainerImageTag
type AddonsVersion map[Addon]*AddonVersion

type KubernetesVersion struct {
	ControlPlaneHostVersion      ControlPlaneHostVersion
	ControlPlaneContainerVersion ControlPlaneContainerVersion
	AddonsVersion                AddonsVersion
}

type KubernetesVersions map[string]KubernetesVersion

var supportVersions = KubernetesVersions{
		"1.15.2": KubernetesVersion{
			ControlPlaneHostVersion: ControlPlaneHostVersion{
				KubeletVersion:          "1.15.2",
				ContainerRuntimeVersion: "1.15.0",
			},
			ControlPlaneContainerVersion: ControlPlaneContainerVersion{
				Hyperkube: &ContainerImageTag{"hyperkube", "v1.15.2"},
				Etcd:      &ContainerImageTag{"etcd", "3.3.11"},
				CoreDNS:   &ContainerImageTag{"coredns", "1.3.1"},
				Pause:     &ContainerImageTag{"pause", "3.1"},
				Tooling:   &ContainerImageTag{"skuba-tooking", "0.1.0"},
			},
			AddonsVersion: AddonsVersion{
				Cilium:  &AddonVersion{"1.5.3", 1},
				Kured:   &AddonVersion{"1.2.0", 0},
				Dex:     &AddonVersion{"2.16.0", 3},
				Gangway: &AddonVersion{"3.1.0-rev4", 3},
				PSP:     &AddonVersion{"", 0},
			},
		},
	}

func ComponentContainerImageWithAvailableVersions(component Component, clusterVersion *version.Version, availableVersions KubernetesVersions) string {
	currentKubernetesVersion := availableVersions[clusterVersion.String()]
	if componentVersion, found := currentKubernetesVersion.ControlPlaneContainerVersion[component]; found {
		return images.GetGenericImage(skuba.ImageRepository, componentVersion.Name, componentVersion.Tag)
	}
	return ""
}
```

Reply:
> Because we have a container image called skuba-tooling, the variable cannot be skuba-tooling :)

Gotcha :sweat_smile: 

> You are right, this is my general thought before, maybe it's better than current.
In this way, we could cover one case that the control plane container image name might be changed in the future. Also, no need to manually add a new component to the switch case.
WDYT?

LGTM

Reply:
OK, I think I finally managed to understand the change here, I like the new map approach :+1:

Reply:
Couldn't we use the same logic for all of the `versions` if we just make all of the `ContainerImageTag` types?

Reply:
Well, cilium addon is much complicated to proceed because it has _cilium-init_, _cilium-operator_, and _cilium_. I am still thinking a way to simplify the code and make it readable and easier to maintain/developing.